lang/java/avro/src/main/java/org/apache/avro/LogicalTypes.java

```diff
@@ -51,9 +65,27 @@ default String getTypeName() {
   public static void register(String logicalTypeName, LogicalTypeFactory factory) {
     Objects.requireNonNull(logicalTypeName, "Logical type name cannot be null");
     Objects.requireNonNull(factory, "Logical type factory cannot be null");
+
+    try {
+      String factoryTypeName = factory.getTypeName();
+      if (!logicalTypeName.equals(factoryTypeName)) {
+        throw new IllegalArgumentException(String.format(
```

Comment:
Changed to a log message. Might be useful if this can be cleaned up at some point, as seems could be confusing that TypeName can be defined in the Factory, but if a different name is used to Register, that will be used instead.

Reply:
Just in case if there is a difference, updated the Specific Compiler to use the method that takes the string key and the factory when generating code.

Technically shouldn't be an issue, as to use a custom logic type factory in the maven plugin, it requires the type name. But seems safer this way for now to me